src/strings/ops.c

```diff
@@ -216,9 +217,50 @@ MVMint64 MVM_string_index(MVMThreadContext *tc, MVMString *haystack, MVMString *
     if (ngraphs > hgraphs || ngraphs < 1)
         return -1;
 
+    /* Fast paths when storage types are identical. Uses memmem function, which
+     * in glibc uses Knuth-Morris-Pratt algorithm as of glibc-2.8-44-g0caca71ac9 */
+    switch (haystack->body.storage_type) {
+        case MVM_STRING_GRAPHEME_32:
+            if (needle->body.storage_type == MVM_STRING_GRAPHEME_32) {
+                /* Keep as void* to not lose precision */
+                void *mm_return_32 = memmem(
+                    haystack->body.storage.blob_32 + start, /* start position */
+                    (hgraphs - start) * sizeof(MVMGrapheme32), /* length of haystack from start position to end */
+                    needle->body.storage.blob_32, /* needle start */
+                    ngraphs * sizeof(MVMGrapheme32) /* needle length */
+                );
+                if (mm_return_32 == NULL)
+                    return -1;
+                /* If we aren't on a 32 bit boundary then freak out!!! */
+                else if (
+                    ( (intptr_t)mm_return_32 - (intptr_t)haystack->body.storage.blob_32) % sizeof(MVMGrapheme32)
```

Comment:
If somebody could check the sanity of this line for determining if the void * pointer returned by memmem() is not on the 32bit integer boundary. Want this double checked.